packages/drawer/src/ux-modal-service.ts

```diff
@@ -0,0 +1,277 @@
+  dismiss?: () => void;
+  ok?: (event: Event) => void;
+}
+
+interface ModalLayer {
+  bindingContext?: ModalBindingContext;
+  drawer: UxDrawer;
+}
+
+@inject(TemplatingEngine, EventAggregator, DefaultDrawerConfiguration)
+export class ModalService {
+
+  public startingZIndex: number = 200;
+  public modalLayers: Array<ModalLayer> = [];
+  public drawerIndex: number = 0;
+
+  constructor(
+    private templatingEngine: TemplatingEngine, 
+    private eventAggregator: EventAggregator,
+    private defaultConfig: DefaultDrawerConfiguration) {
+
+  }
+
+  public addLayer(drawer: UxDrawer, bindingContext: ModalBindingContext) {
+    this.modalLayers.push({
+      bindingContext: bindingContext,
+      drawer: drawer
+    });
+    if (this.modalLayers.length === 1) {
+      this.setKeyListener();
+    }
+  }
+
+  public getLayer(drawer: UxDrawer): ModalLayer | undefined {
+    const index = this.modalLayers.map(i => i.drawer).indexOf(drawer);
+    if (index !== -1) {
+      return this.modalLayers[index];
+    }
+    return undefined;
+  }
+
+  public removeLayer(drawer: UxDrawer) {
+    const index = this.modalLayers.map(i => i.drawer).indexOf(drawer);
+    if (index !== -1) {
+      this.modalLayers.splice(index, 1);
+    }
+    if (this.modalLayers.length === 0) {
+      this.removeKeyListener();
+    }
+  }
+
+  private setKeyListener() {
+    window.addEventListener('keyup', this);
+  }
+
+  private removeKeyListener() {
+    window.removeEventListener('keyup', this);
+  }
+
+  public handleEvent(event: KeyboardEvent) {
+    const key = this.getActionKey(event);
+    if (key === undefined) { return; }
+    const activeLayer = this.getLastDrawer();
+    const keyboard = activeLayer.keyboard;
+    if (key === 'Escape'
+      && (keyboard === true || keyboard === key || (Array.isArray(keyboard) && keyboard.indexOf(key) > -1))) {
+      activeLayer.dismiss();
+    } else if (key === 'Enter' && (keyboard === key || (Array.isArray(keyboard) && keyboard.indexOf(key) > -1))) {
+      activeLayer.ok();
+    }
+  }
+
+  private getActionKey(event: KeyboardEvent): DrawerKeybord | undefined {
+    if ((event.code || event.key) === 'Escape' || event.keyCode === 27) {
+      return 'Escape';
+    }
+    if ((event.code || event.key) === 'Enter' || event.keyCode === 13) {
+      return 'Enter';
+    }
+    return undefined;
+  }
+
+  public getLastLayer(): ModalLayer {
+    return this.modalLayers[this.modalLayers.length - 1];
+  }
+
+  public getLastDrawer(): UxDrawer {
+    return this.getLastLayer().drawer;
+  }
+
+  public get zIndex() {
+    return this.startingZIndex + this.modalLayers.length;
+  }
+
+  public open(options: ModalServiceOptions): UxDrawer & ModalServiceDrawer {
+    // const defaultConfig = this.container.get(DefaultDrawerConfiguration);
+    options = Object.assign({}, this.defaultConfig, options);
+    this.drawerIndex++;
+    const bindingContext: ModalBindingContext = {};
+    if (!options.viewModel && !options.view) {
```

Comment:
maybe doing this check right after normalizing `options` in line 140, it's better and can help avoid 1 mutation of `drawerIndex`, less surprises during debugging